pkg/storegateway/storepb/custom.go

```diff
@@ -390,3 +391,78 @@ func (x LabelMatcher_Type) PromString() string {
 func (m *Series) PromLabels() labels.Labels {
 	return mimirpb.FromLabelAdaptersToLabels(m.Labels)
 }
+
+type PreallocatingSliceMetric struct {
+	mimirpb.Metric
+}
+
+func (m *PreallocatingSliceMetric) Unmarshal(dAtA []byte) error {
```

Comment:
I really feel we need to modify protogen or something to start autogenerating this. We're having more and more custom-modified autogenerated code :(

Reply:
Meanwhile, I'd write a comment here explaining what you're doing, so the readers wouldn't need to read the code.

Reply:
added a comment in 158d167276642e6ed3fcdb258c6cc8c48e6b077d

i still haven't investigated whether we can reuse the code generation that gogoproto does and only change it a bit instead of having to also copy the code generation code

I don't know how similar to this use-case other modifications of the generated code are - i.e. can we unify them with a generator or will we just write N different generators for the other N cases.

I know mimirpb.LabelAdaptor is a different use-case.

Reply:
for `PreallocWriteRequest` and `PreallocTimeseries`, which seem to only be using pooling we can maybe use [vtprotobuf](https://github.com/planetscale/vtprotobuf) - that has pooling generation

Reply:
I would move all this function logic just used to count the number of labels into a dedicated (non exported) function. Then this unmarshal would be trivial to read because it's just about pre-allocating the slice and calling the "parent".